src/org/rascalmpl/library/lang/rascal/tests/concrete/recovery/ParseErrorDisambiguator.java

```diff
@@ -0,0 +1,174 @@
+package org.rascalmpl.library.lang.rascal.tests.concrete.recovery;
+
+import org.rascalmpl.interpreter.asserts.Ambiguous;
+import org.rascalmpl.values.IRascalValueFactory;
+import org.rascalmpl.values.RascalValueFactory;
+import org.rascalmpl.values.parsetrees.ITree;
+import org.rascalmpl.values.parsetrees.ProductionAdapter;
+import org.rascalmpl.values.parsetrees.TreeAdapter;
+
+import io.usethesource.vallang.IBool;
+import io.usethesource.vallang.IConstructor;
+import io.usethesource.vallang.IList;
+import io.usethesource.vallang.IListWriter;
+import io.usethesource.vallang.ISet;
+import io.usethesource.vallang.ISetWriter;
+import io.usethesource.vallang.IValue;
+import io.usethesource.vallang.type.Type;
+
+public class ParseErrorDisambiguator {
+	private final IRascalValueFactory rascalValues;
+    private boolean allowAmbiguity;
+
+    public ParseErrorDisambiguator(IRascalValueFactory rascalValues) {
+		super();
+		
+		this.rascalValues = rascalValues;
+    }
+
+    private static class ScoredTree {
+        public final IConstructor tree;
+        public final int score;
+
+        public ScoredTree(IConstructor tree, int score) {
+            this.tree = tree;
+            this.score = score;
+        }
+    }
+
+    /**
+     * Disambiguate error trees, faster implementation of the original Rascal variant:
+Tree defaultErrorDisambiguationFilter(Tree t) {
+  return visit(t) {
+    case a:amb(_) => ambDisambiguation(a)
+  };
+}
+
+private Tree ambDisambiguation(amb(set[Tree] alternatives)) {
+  // Go depth-first
+  rel[int score, Tree alt] scoredErrorTrees = { <scoreErrors(alt), alt> | Tree alt <- alternatives };
+  set[Tree] nonErrorTrees = scoredErrorTrees[0];
+
+  if (nonErrorTrees == {}) {
+    return (getFirstFrom(scoredErrorTrees) | it.score > c.score ? c : it | c <- scoredErrorTrees).alt;
+  }
+  
+  if ({Tree single} := nonErrorTrees) {
+    // One ambiguity left, no ambiguity concerns here
+    return single;
+  }
+
+  // Multiple non-error trees left, return an ambiguity node with just the non-error trees
+  return amb(nonErrorTrees);
+}
+
+private int scoreErrors(Tree t) = (0 | it + getSkipped(e).src.length | /e:appl(error(_,_,_),_) := t);
+
+// Handle char and cycle nodes
+default Tree defaultErrorDisambiguationFilter(Tree t) = t;
+     **/
+
+    public synchronized IConstructor disambiguateErrors(IConstructor arg, IBool allowAmbiguity) {
+        this.allowAmbiguity = allowAmbiguity != null && allowAmbiguity.getValue();
+        return disambiguate(arg).tree;
+    }
+
+    private ScoredTree disambiguate(IConstructor tree) {
+        Type type = tree.getConstructorType();
+
+		if (type == RascalValueFactory.Tree_Appl) {
+            return disambiguateAppl((ITree) tree);
+        } else if (type == RascalValueFactory.Tree_Amb) {
+            return disambiguateAmb((ITree) tree);
+        }
```

Comment:
are the other tree types not relevant? If so, document here why?

since the rascal visit would also go into all the other subtrees (like cycle) and still match appl & ambs in there)

what if error tree is at the top?

Reply:
No other tree types can have trees as children. Cycle is just a simple node:
```
data Tree //(loc src = |unknown:///|(0,0,<0,0>,<0,0>))
     = appl(Production prod, list[Tree] args) // <1>
     | cycle(Symbol symbol, int cycleLength)  // <2>
     | amb(set[Tree] alternatives) // <3>
     | char(int character) // <4>
     ;
```
An error node at the top would not be an issue as it is just a regular appl and handled as such.